onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/data_types.h"
+#include "core/framework/sparse_tensor.h"
+
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+
+SparseTensor::SparseTensor(MLDataType elt_type,
+                           const TensorShape& shape,
+                           size_t nnz,
+                           void* values_data,
+                           void* indices_data,
+                           const OrtAllocatorInfo& allocator_info)
+    : values_(elt_type, TensorShape({static_cast<int64_t>(nnz)}), values_data, allocator_info, 0),
```

Comment:
>static_cast<int64_t> [](start = 37, length = 20)

Is there a reason to use size_t for nnz instead of int64_t

Reply:
Well, I have been puzzling about this. Is the question about signed vs unsigned or size_t vs. uint64_t? It seems to me that an unsigned type is better than signed. It seems to me that even TensorShape would be better off using unsigned types. I suspect that the use of signed values in TensorShape originates from frameworks (like Caffe2) which use negative values to denote an unknown dimension, but we use TensorShape only for known shapes (that is, runtime shapes). So, in general, we seem to run into these kind of casts in many places, and the ideal solution may require changes elsewhere. But, for now, if int64_t is better here, I can change it.

Reply:
> but we use TensorShape only for known shapes
It's not true.

Reply:
Where do we use TensorShape for an unknown shape? (As opposed to ONNX's TensorShapeProto.)

Reply:
See: https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/framework/tensor_shape.h#L87

Reply:
That's the API documentation. I am curious where it is used. In my view, there are two different categories of shapes: one is a "static shape", where we may have unknown dimensions, and the other is a "dynamic or runtime shape" where all dimensions are known. There is value in distinguishing these two concepts. In ONNX, there is a distinction: we use TensorShapeProto which uses strings like "N" to denote unknown dimensions and integers to denote known dimension values; on the other hand TensorProto has a specific tensor value, so its shape is a repeated sequence of integer values (which should not have an unknown value). 

There is a third category: in ops like reshape, we use a -1 in the desired output-shape to indicate that a particular dimension value should be computed from the remaining dimensions and total size. Ideally, we should not be using a TensorShape there (the shape is represented as a Tensor there anyway). 

So, I wonder if there is any place where we do use negative values in a TensorShape.

Reply:
There are a couple of places I know of, but they could be changed. In If.cc and scan_utils.cc I converted the TensorShapeProto from the NodeArg for a subgraph output into TensorShape using onnxruntime::utils::GetTensorShapeFromTensorShapeProto, but that was largely for convenience to be able to check if the shape was concrete or not (allocation of the output is delayed until the concrete shape is known so the complete output can be allocated at once at the control flow node level). That code could use the TensorShapeProto just as easily if we added a helper to check that for a symbolic dimension. 

ExecutionFrame::AllocateMLValueTensorSelfOwnBufferHelper checks that the TensorShape has no symbolic dims so by the time we want to use the shape for allocation it must be concrete. Not sure if there's any period where TensorShape needs to support a negative dimension though. Probably need to throw if it has one and see if any tests break due to that.